Fix PSBT deserialization of 0-input transactions

achow101 opened the pull request:
0-input transactions can be ambiguously deserialized as being witness transactions. Since the unsigned transaction is never serialized as a witness transaction as it has no witnesses, we should always deserialize it as a non-witness transaction and set the serialization flags as such.

When a transaction is serliazed for the non-witness-utxo, it is always a valid network transaction and thus it should be always be deserialized as a witness transaction and the deserialzation flags are set as such.

Fixes #13958

sipa commented on src/script/sign.h:
```diff
@@ -302,6 +302,9 @@ struct PSBTInput
                     } else if (key.size() != 1) {
                         throw std::ios_base::failure("Non-witness utxo key is more than one byte type");
                     }
+                    // Set the stream to unserialize with witness since this is always a valid network transaction
```
You can instead use `UnserializeFromVector(OverrideStream(&s, s.GetType(), s.GetVersion() & ~SERIALIZE_TRANSACTION_NO_WITNESS);` to avoid permanently changing the flags on `s` (similar below).

achow101 replied:
Done

sipa commented:
Would you mind also just dropping the witness after reading the non-witness UTXO field (and/or before writing)?

achow101 committed:
Fix PSBT deserialization of 0-input transactions

0-input transactions can be ambiguously deserialized as being witness
transactions. Since the unsigned transaction is never serialized as
a witness transaction as it has no witnesses, we should always
deserialize it as a non-witness transaction and set the serialization
flags as such.

Also always serialize the unsigned transaction as a non-witness transaction.

achow101 committed:
Serialize non-witness utxo as a non-witness tx but always deserialize as witness

Strip out the witnesses when serializing the non-witness utxo. However
witness serializations are allowed, so make sure we always deserialize
as witness.

achow101 commented:
> Would you mind also just dropping the witness after reading the non-witness UTXO field (and/or before writing)?

Done.

sipa commented:
utACK bd19cc78cfc455cf06e120adb0d12c2f96ba8fca

Empact commented on src/script/sign.h:
```diff
@@ -223,7 +223,8 @@ struct PSBTInput
         if (non_witness_utxo) {
             SerializeToVector(s, PSBT_IN_NON_WITNESS_UTXO);
-            SerializeToVector(s, non_witness_utxo);
+            OverrideStream<Stream> os(&s, s.GetType(), s.GetVersion() | SERIALIZE_TRANSACTION_NO_WITNESS);
```
nit: could use the `WithOrVersion` helper here:
https://github.com/bitcoin/bitcoin/blob/ddc3ec92b0e655a3da21ac2e85ec2e7ecb66c65b/src/streams.h#L66-L70

Empact commented on src/streams.h:
```diff
@@ -61,6 +61,7 @@ class OverrideStream
 
     int GetVersion() const { return nVersion; }
     int GetType() const { return nType; }
+    size_t size() const { return stream->size(); }
```
nit: could be nice to put `size` with `read`, `write`, as they are all delegated.

laanwj commented:
utACK bd19cc78cfc455cf06e120adb0d12c2f96ba8fca

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/13960/commits/bd19cc78cfc455cf06e120adb0d12c2f96ba8fca

fanquake commented:
Will be backported in #13976

dongcarl commented:
> When a transaction is serliazed for the non-witness-utxo, it is always a valid network transaction and thus it should be always be deserialized as a witness transaction and the deserialzation flags are set as such.

Surely you mean "...it should be always be deserialized as a _**non-witness**_ transaction..."

sipa commented:
@dongcarl No. "Deserialized as a witness transaction" here means "permit witnesses, and if ambiguous, prefer the interpretation with witness over the one with 0 inputs". As it's always a valid fully-signed transaction, it can't have 0 inputs, and thus we should indeed prefer the witness version.

dongcarl commented:
@sipa Makes sense!
